DOCS-5309 Istio Integration Edits (#15231)
DataDog/integrations-core: istio/README.md

## Overview

Datadog monitors every aspect of your Istio environment, so you can:
- Assess the health of Envoy and the Istio control plane with [logs](#log-collection).
- Break down the performance of your service mesh with [request, bandwidth, and resource consumption metrics](#metrics).
- Map network communication between containers, pods, and services over the mesh with [Network Performance Monitoring][1].
- Drill into distributed traces for applications transacting over the mesh with [APM][2].

To learn more about monitoring your Istio environment with Datadog, [see the Monitor blog post][3].

## Setup

Follow the instructions below to install and configure this check for an Agent running on a host. For containerized environments, see the [Autodiscovery Integration Templates][4] for guidance on applying these instructions.

This OpenMetrics-based integration has a latest mode (`use_openmetrics`: true) and a legacy mode (`use_openmetrics`: false). To get all the most up-to-date features, Datadog recommends enabling the latest mode. For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][25].

> 
> When you enable the `use_openmetrics` configuration option, the Istio integration uses the latest version of the OpenMetrics-based implementation of the check. 

Reviewer: ```suggestion
```

This looks redundant with the previous paragraph.

> If you have multiple instances of Datadog collecting Istio metrics, make sure you are using the same version for all of them. Otherwise, metrics data may fluctuate on the Datadog site.

Reviewer: ```suggestion
If you have multiple instances of Datadog collecting Istio metrics, make sure you are using the same mode for all of them. Otherwise, metrics data may fluctuate on the Datadog site.
```

> Metrics marked as `[OpenMetrics V1]`, `[OpenMetrics V2]`, or `[OpenMetrics V1 and V2]` are only available using the corresponding version of the Istio integration. Metrics marked as `Istio v1.5+` are collected using Istio version 1.5 or later.

Reviewer: ```suggestion
Metrics marked as `[OpenMetrics V1]`, `[OpenMetrics V2]`, or `[OpenMetrics V1 and V2]` are only available using the corresponding mode of the Istio integration. Metrics marked as `Istio v1.5+` are collected using Istio version 1.5 or later.
```

## Overview

Datadog monitors every aspect of your Istio environment, so you can:
- Assess the health of Envoy and the Istio control plane with [logs](#log-collection).
- Break down the performance of your service mesh with [request, bandwidth, and resource consumption metrics](#metrics).
- Map network communication between containers, pods, and services over the mesh with [Network Performance Monitoring][1].
- Drill into distributed traces for applications transacting over the mesh with [APM][2].

To learn more about monitoring your Istio environment with Datadog, [see the Monitor blog post][3].

## Setup

Follow the instructions below to install and configure this check for an Agent running on a host. For containerized environments, see the [Autodiscovery Integration Templates][4] for guidance on applying these instructions.

> ### Prerequisites
> 
> This OpenMetrics-based integration has a latest version (V2) and a legacy version (V1). To get all the most up-to-date features, Datadog recommends upgrading to the latest version. For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][25].

Author: ```suggestion
This OpenMetrics-based integration has a latest mode (`use_openmetrics`: true) and a legacy mode (`use_openmetrics`: false). To get all the most up-to-date features, Datadog recommends enabling the latest mode. For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][25].
```

> This OpenMetrics-based integration has a latest version (V2) and a legacy version (V1). To get all the most up-to-date features, Datadog recommends upgrading to the latest version. For more information, see the [Latest and Legacy Versioning For OpenMetrics-based Integrations][25].

Author: ```suggestion
This OpenMetrics-based integration has a latest version (V2) and a legacy version (V1). To get all the most up-to-date features, Datadog recommends upgrading to the latest version. For more information, see [Latest and Legacy Versioning For OpenMetrics-based Integrations][25].
```

> When you enable the `use_openmetrics` configuration option, the Istio integration uses the OpenMetrics V2 implementation of the check. In OpenMetrics V2, metrics are submitted more accurately by default and behave closer to Prometheus metric types. For example, Prometheus metrics ending in  `_count` and `_sum` are submitted as `monotonic_count` by default.

Reviewer: Should we replace "OpenMetrics V2" with "latest" in this paragraph? To reduce the proliferation of terms used to refer to it. Also, I think that we should be able to remove the second sentence of this paragraph because that's common to all integrations with the V2 / V1 divide, and if anything we should probably consider adding this to the [new guide on the topic](https://github.com/DataDog/documentation/blob/master/content/en/integrations/guide/versions-for-openmetrics-based-integrations.md).

Author: ```suggestion
When you enable the `use_openmetrics` configuration option, the Istio integration uses the OpenMetrics implementation of the check. In OpenMetrics, metrics are submitted more accurately by default and behave closer to Prometheus metric types. For example, Prometheus metrics ending in  `_count` and `_sum` are submitted as `monotonic_count` by default.
```

## Troubleshooting

### Invalid chunk length error

> If you see the following error on the OpenMetrics-based implementation of Istio (Istio integration version `3.13.0` or earlier):

Reviewer: ```suggestion
If you see the following error on the legacy mode of the Istio integration (Istio integration version `3.13.0` or earlier):
```

```python
  Error: ("Connection broken: InvalidChunkLength(got length b'', 0 bytes read)",
  InvalidChunkLength(got length b'', 0 bytes read))
```

> You can use the latest version of the OpenMetrics-based Istio integration to resolve this error.

Reviewer: ```suggestion
You can use the *latest* mode of the OpenMetrics-based Istio integration to resolve this error.
```

You must upgrade to at minimum Agent `7.31.0` and Python 3. See the [Configuration](#configuration) section to enable OpenMetrics.

### Using the generic OpenMetrics integration in an Istio deployment

If Istio proxy sidecar injection is enabled, monitoring other Prometheus metrics using the [OpenMetrics integration][20] with the same metrics endpoint as `istio_mesh_endpoint` can result in high custom metrics usage and duplicated metric collection.

To ensure that your OpenMetrics configuration does not redundantly collect metrics, either:

1. Use specific metric matching in the `metrics` configuration option, or
2. If using the wildcard `*` value for `metrics`, consider using the following OpenMetrics integration options to exclude metrics already supported by the Istio and Envoy integrations.

> #### OpenMetrics V2 configuration with generic metric collection

Reviewer: ```suggestion
#### OpenMetrics *latest* mode configuration with generic metric collection
```

## Troubleshooting

### Invalid chunk length error
If you see the following error on OpenMetricsBaseCheck (V1) implementation of Istio (Istio integration version `3.13.0` or earlier):

```python
  Error: ("Connection broken: InvalidChunkLength(got length b'', 0 bytes read)",
  InvalidChunkLength(got length b'', 0 bytes read))
```

> You can use the OpenMetrics V2 implementation of the Istio integration to resolve this error.

Author: ```suggestion
You can use the latest version of the OpenMetrics-based Istio integration to resolve this error.
```

> You must upgrade to at minimum Agent `7.31.0` and Python 3. See the [Configuration](#configuration) section to enable OpenMetrics V2.

Author: ```suggestion
You must upgrade to at minimum Agent `7.31.0` and Python 3. See the [Configuration](#configuration) section to enable OpenMetrics.
```

### Using the generic OpenMetrics integration in an Istio deployment

If Istio proxy sidecar injection is enabled, monitoring other Prometheus metrics using the [OpenMetrics integration][20] with the same metrics endpoint as `istio_mesh_endpoint` can result in high custom metrics usage and duplicated metric collection.

To ensure that your OpenMetrics configuration does not redundantly collect metrics, either:

1. Use specific metric matching in the `metrics` configuration option, or
2. If using the wildcard `*` value for `metrics`, consider using the following OpenMetrics integration options to exclude metrics already supported by the Istio and Envoy integrations.

#### OpenMetrics V2 configuration with generic metric collection

> Be sure to exclude Istio and Envoy metrics from your configuration to avoid high custom metrics billing. Use `exclude_metrics` if using the OpenMetrics V2 configuration (`openmetrics_endpoint` enabled).

Author: ```suggestion
Be sure to exclude Istio and Envoy metrics from your configuration to avoid high custom metrics billing. Use `exclude_metrics` if `openmetrics_endpoint` is enabled.
```

```

> #### OpenMetrics Legacy configuration with generic metric collection

Author: ```suggestion
#### OpenMetrics legacy configuration with generic metric collection
```

```

> #### OpenMetrics V1 configuration (Legacy) with generic metric collection

Author: ```suggestion
#### OpenMetrics Legacy configuration with generic metric collection
```

> Be sure to exclude Istio and Envoy metrics from your configuration to avoid high custom metrics billing. Use `ignore_metrics` if using the OpenMetrics V1 configuration (`prometheus_url` enabled).

Author: ```suggestion
Be sure to exclude Istio and Envoy metrics from your configuration to avoid high custom metrics billing. Use `ignore_metrics` if `prometheus_url` is enabled.
```
